File: gesttravaux-symfony/src/Entity/Prestation.php
<?php

namespace App\Entity;

use App\Repository\PrestationRepository;
use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;
use Doctrine\ORM\Mapping as ORM;

/**
 * Prestation entity - Service types
 *
 * Represents specific types of services offered within a category
 * (e.g., "Installation de chaudière" in category "Plumbing")
 *
 * @author GestTravaux Pro
 */
#[ORM\Entity(repositoryClass: PrestationRepository::class)]
#[ORM\Table(name: 'prestations')]
class Prestation
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column(name: 'idPrestation')]
    private ?int $id = null;

    #[ORM\Column(length: 45)]
    private ?string $libelle = null;

    #[ORM\ManyToOne(targetEntity: Categorie::class, inversedBy: 'prestations')]
    #[ORM\JoinColumn(name: 'categories_idCategorie', referencedColumnName: 'idCategorie', nullable: false)]
    private ?Categorie $categorie = null;

    /**
     * @var Collection<int, Devis>
     */
    #[ORM\OneToMany(targetEntity: Devis::class, mappedBy: 'prestation')]
    private Collection $devis;

    public function __construct()
    {
        $this->devis = new ArrayCollection();
    }

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getLibelle(): ?string
    {
        return $this->libelle;
    }

    public function setLibelle(string $libelle): static
    {
        $this->libelle = $libelle;

        return $this;
    }

    public function getCategorie(): ?Categorie
    {
        return $this->categorie;
    }

    public function setCategorie(?Categorie $categorie): static
    {
        $this->categorie = $categorie;

        return $this;
    }

    /**
     * @return Collection<int, Devis>
     */
    public function getDevis(): Collection
    {
        return $this->devis;
    }

    public function addDevi(Devis $devi): static
    {
        if (!$this->devis->contains($devi)) {
            $this->devis->add($devi);
            $devi->setPrestation($this);
        }

        return $this;
    }

    public function removeDevi(Devis $devi): static
    {
        if ($this->devis->removeElement($devi)) {
            // set the owning side to null (unless already changed)
            if ($devi->getPrestation() === $this) {
                $devi->setPrestation(null);
            }
        }

        return $this;
    }

    public function __toString(): string
    {
        return $this->libelle ?? '';
    }
}

File: gesttravaux-java/src/main/java/com/immosync/gesttravaux/models/Prestation.java
package com.immosync.gesttravaux.models;

import jakarta.persistence.*;
import java.util.HashSet;
import java.util.Set;

/**
 * Entity representing a service/prestation.
 * Each prestation belongs to one category and can be included in multiple devis.
 *
 * @author GestTravaux Pro
 * @version 1.0
 */
@Entity
@Table(name = "PRESTATIONS")
public class Prestation {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "idPrestation")
    private Integer idPrestation;

    @Column(name = "libelle", length = 45, nullable = false)
    private String libelle;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "CATEGORIES_idCategorie", nullable = false)
    private Categorie categorie;

    @OneToMany(mappedBy = "prestation", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    private Set<Devis> devisList = new HashSet<>();

    /**
     * Default constructor required by JPA.
     */
    public Prestation() {
    }

    /**
     * Constructor with essential fields.
     *
     * @param libelle prestation label/name
     * @param categorie prestation category
     */
    public Prestation(String libelle, Categorie categorie) {
        this.libelle = libelle;
        this.categorie = categorie;
    }

    // Getters and Setters

    /**
     * Gets the prestation ID.
     *
     * @return the prestation ID
     */
    public Integer getIdPrestation() {
        return idPrestation;
    }

    /**
     * Sets the prestation ID.
     *
     * @param idPrestation the prestation ID
     */
    public void setIdPrestation(Integer idPrestation) {
        this.idPrestation = idPrestation;
    }

    /**
     * Gets the prestation label.
     *
     * @return the label
     */
    public String getLibelle() {
        return libelle;
    }

    /**
     * Sets the prestation label.
     *
     * @param libelle the label
     */
    public void setLibelle(String libelle) {
        this.libelle = libelle;
    }

    /**
     * Gets the prestation category.
     *
     * @return the category
     */
    public Categorie getCategorie() {
        return categorie;
    }

    /**
     * Sets the prestation category.
     *
     * @param categorie the category
     */
    public void setCategorie(Categorie categorie) {
        this.categorie = categorie;
    }

    /**
     * Gets the list of devis for this prestation.
     *
     * @return set of devis
     */
    public Set<Devis> getDevisList() {
        return devisList;
    }

    /**
     * Sets the list of devis for this prestation.
     *
     * @param devisList set of devis
     */
    public void setDevisList(Set<Devis> devisList) {
        this.devisList = devisList;
    }

    @Override
    public String toString() {
        return libelle;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Prestation)) return false;
        Prestation that = (Prestation) o;
        return idPrestation != null && idPrestation.equals(that.getIdPrestation());
    }

    @Override
    public int hashCode() {
        return getClass().hashCode();
    }
}
